### ISSUE TITLE
cardano-sl hangs after losing connection with the Internet

### ISSUE THREAD

**Issue by zmeel** [input-output-hk/cardano-sl#2269]
I'm running the cardano-sl node in a Docker container 24/7 on a Synology DS415+ (8 Gb RAM). My modem is on a programmable clock and every other day the modem is turned off and on (at 03:00) to have it restart. Time between off and on is 1 minute plus the modem needs to restart so in total connection to the Internet is lost for about 3 minutes. Two days ago I have written a Python script that scrapes the log file for some data and I plot this data in Grafana. Everything looked nice and I could see that on average 3 blocks were created every minute. However, the next morning I saw no blocks were created after 03:00. The node seemed OK and also Daedalus was not reporting any issue so I restarted the node and this morning again no new blocks created after 03:00 while everything seemed fine:

This is a snippet from the log at 03:00
<pre>
[94m[node:INFO:ThreadId 447402] [0m[2018-01-07 03:00:00 UTC] Block has been adopted 683cd8f3
[92m[node:DEBUG:ThreadId 447399] [0m[2018-01-07 03:00:00 UTC] self: sent MsgRequestBlocks (fromList [NodeId 13.250.124.35:3000:0]) to NodeId 13.250.124.35:3000:0
[92m[node.worker.block:DEBUG:ThreadId 560] [0m[2018-01-07 03:00:00 UTC] Waiting on the block queue or recovery header var
[94m[node.wallet.servant:INFO:ThreadId 447441] [0m[2018-01-07 03:00:01 UTC] 
[36mGET[0m
    [37m:>[0m api
    [37m:>[0m settings
    [37m:>[0m sync
    [37m:>[0m progress
   [37mStatus:[0m [32mOK[0m 0.001876247s[37m > [0mprogress=454487/454487 peers=0
[94m[node.wallet.servant:INFO:ThreadId 447444] [0m[2018-01-07 03:00:01 UTC] 
</pre>

And this is from almost 2 hours later:
<pre>
[92m[node.worker.block:DEBUG:ThreadId 495216] [0m[2018-01-07 04:59:11 UTC] After we took lock for genesis block creation, we noticed that we shouldn't create it
92m[node.worker.block:DEBUG:ThreadId 495216] [0m[2018-01-07 04:59:11 UTC] MemPool metrics release: modify time was 3018mcs size is 0
[92m[node.worker.slotting.slotting:DEBUG:ThreadId 566] [0m[2018-01-07 04:59:11 UTC] Waiting for 19986409mcs before new slot
[92m[node.worker.block:DEBUG:ThreadId 495215] [0m[2018-01-07 04:59:11 UTC] Block with depth 'k' (2160) was created during slot 20384th slot of 20th epoch
[92m[node.worker.block:DEBUG:ThreadId 495215] [0m[2018-01-07 04:59:11 UTC] Overall chain quality is 0.999
[92m[node.worker.block:DEBUG:ThreadId 495216] [0m[2018-01-07 04:59:11 UTC] Our pk: pub:71d8487a, our pkHash: d45ee6dc
[92m[node.worker.block:DEBUG:ThreadId 495216] [0m[2018-01-07 04:59:11 UTC] Current slot leader: 5071d880
[92m[node.worker.block:DEBUG:ThreadId 495216] [0m[2018-01-07 04:59:11 UTC] Trimmed leaders: [(1293, 5071d880), (1294, 1deb8290), (1295, 43011479), (1296, 43011479), (1297, 43011479), (1298, 65904a89), (1299, af2800c1), (1300, 6c9e1497), (1301, 6c9e1497), (1302, af2800c1), (1303, 5071d880), (1304, af2800c1), (1305, 6c9e1497), (1306, 5411c7bf), (1307, 65904a89), (1308, 5411c7bf), (1309, 65904a89), (1310, 43011479), (1311, 6c9e1497), (1312, 65904a89), (1313, 1deb8290)]
[92m[node.worker.block:DEBUG:ThreadId 495216] [0m[2018-01-07 04:59:11 UTC] Available to use lightweight PSKs: []
[92m[node.worker.block:DEBUG:ThreadId 495216] [0m[2018-01-07 04:59:11 UTC] End delegation psk for this slot: ProxySk { w = epoch #0, iPk = pub:993a8f05, dPk = pub:89c29f8c }
[94m[node.wallet.servant:INFO:ThreadId 495218] [0m[2018-01-07 04:59:11 UTC] 
[36mGET[0m
    [37m:>[0m api
    [37m:>[0m settings
    [37m:>[0m sync
    [37m:>[0m progress
   [37mStatus:[0m [32mOK[0m 0.001877419s[37m > [0mprogress=454487/454487 peers=0
[94m[node.wallet.servant:INFO:ThreadId 495221] [0m[2018-01-07 04:59:11 UTC] 
</pre>

The node is still at block 454487 two hours after the Internet outage. A restart of the node the log shows:
<pre>
[94m[node:INFO:ThreadId 4] [0m[2018-01-07 06:58:44 UTC] Our verInfo: VerInfo { magic=2d964a09, blockVersion=0.1.0, inSpecs={96: Conv 43, 49: Conv 5c, 97: Conv 3d, 34: Conv 5e, 98: Conv 37, 67: Conv 60, 83: Conv 0, 4: Conv 5, 5: Conv 4, 37: Conv 5e, 6: Conv 7, 55: Conv 62, 73: Conv 5f, 43: Conv 5d, 92: Conv 31, 61: Conv 61, 93: Conv 2b, 94: Conv 25, 95: Conv 49}, outSpecs={49: Conv 5c, 67: Conv 60, 83: Conv 0, 4: Conv 5, 5: Conv 4, 37: Conv 5e, 6: Conv 7, 55: Conv 62, 73: Conv 5f, 43: Conv 5d, 13: Conv 0, 61: Conv 61}}
[92m[node:DEBUG:ThreadId 4] [0m[2018-01-07 06:58:44 UTC] startNode, we are NodeId 1554e190-5f80-4336-b707-00f53afaf500
[94m[node:INFO:ThreadId 4] [0m[2018-01-07 06:58:44 UTC] cardano-sl: commit unknown
[94m[node:INFO:ThreadId 4] [0m[2018-01-07 06:58:44 UTC] Application: cardano-sl:0, last known block version 0.1.0
[94m[node:INFO:ThreadId 4] [0m[2018-01-07 06:58:44 UTC] My public key is: pub:71d8487a, pk hash: d45ee6dc
[94m[node:INFO:ThreadId 4] [0m[2018-01-07 06:58:44 UTC] Genesis stakeholders (7 addresses, dust threshold 7 coin(s)): GenesisWStakeholders: {1deb8290: 1, 43011479: 1, 5071d880: 1, 5411c7bf: 1, 65904a89: 1, 6c9e1497: 1, af2800c1: 1}
[94m[node:INFO:ThreadId 4] [0m[2018-01-07 06:58:44 UTC] First genesis block hash: 89d9b5a5, genesis seed is 76617361206f7061736120736b6f766f726f64612047677572646120626f726f64612070726f766f6461
</pre>

And then after some more seconds:
<pre>
[92m[node.worker.block:DEBUG:ThreadId 558] [0m[2018-01-07 06:58:58 UTC] Checkpoints available, headers request assembled
[92m[node.worker.block:DEBUG:ThreadId 558] [0m[2018-01-07 06:58:58 UTC] Updating recovery header...
[92m[node.worker.block:DEBUG:ThreadId 558] [0m[2018-01-07 06:58:58 UTC] Recovery started with nodeId="13.250.86.93:3000:0" and tip=4ba6b9b5
[92m[node.worker.block:DEBUG:ThreadId 558] [0m[2018-01-07 06:58:58 UTC] self: inFlight = fromList []
[92m[node.worker.block:DEBUG:ThreadId 558] [0m[2018-01-07 06:58:58 UTC] self: message MsgRequestBlockHeaders (Just (fromList [NodeId 13.250.86.93:3000:0])) enqueued to [NodeId 13.250.86.93:3000:0]
[92m[node:DEBUG:ThreadId 648] [0m[2018-01-07 06:58:58 UTC] self: sending MsgRequestBlockHeaders (Just (fromList [NodeId 13.250.86.93:3000:0])) to NodeId 13.250.86.93:3000:0
[94m[node.wallet.servant:INFO:ThreadId 653] [0m[2018-01-07 06:58:59 UTC] 
[36mGET[0m
    [37m:>[0m api
    [37m:>[0m settings
    [37m:>[0m sync
    [37m:>[0m progress
   [37mStatus:[0m [32mOK[0m 0.002227451s[37m > [0mprogress=454487/455204 peers=0
</pre>

It seems that the node is unaware that the Internet was away for a couple of minutes after 03:00. In the log files between 03:00 and 06:58 (when I restarted the node) `node.worker.block:DEBUG` was not mentioned once. I'm curious if this is normal behavior. Internet outage does occur and Cardano-SL and Daedalus appeared to function normally after an outage although they actually do nothing while they should at least warn they need a restart or even better just pick up the work after an outage...

 I have saved all the log files.

**Comment by zmeel** [comment]
@h4ck3rm1k3: James, I'm hoping you are reading this. 

Normally, when a new issue is opened that developers deem important they react to this issue quite rapidly. Apparently, my issue is not and it could be that the issue is caused by the way I run the node (i.e. in a Docker container). Could you run a simple test to see cardano-sl also hangs in your build if you cut off the Internet for a few minutes? In my case, the node seems totally unaware that the connection with the Internet was gone...

**Comment by h4ck3rm1k3** [comment]
ok I am looking, are you on chat https://chat.cardanohub.org/channel/developers

**Comment by h4ck3rm1k3** [comment]
I second that this is a major issue. I propose to setup a 
man in the middle https://mitmproxy.org/ 
see  cardano-sl/docs/network/topology-configuration.md
it seems that you can define a nat server. Need more time to look into this.

**Comment by zmeel** [comment]
I could NOT confirm this issue on my MacBook Pro. 

I opened Daedalus and let it sync till 100%. Daedalus opens and show my wallet. Then I turned Wifi off. Daedalus does not react and the log shows:

**19:35 Wifi off:**
<pre>
[94m[node.wallet.servant:INFO:ThreadId 2862] [0m[2018-01-08 19:35:13 CET] 
[36mGET[0m
    [37m:>[0m api
    [37m:>[0m settings
    [37m:>[0m sync
    [37m:>[0m progress
   [37mStatus:[0m [32mOK[0m 0.000514s[37m > [0mprogress=461612/461612 peers=0
</pre>

**19:40 Wifi back on:**
<pre>
[94m[node.wallet.servant:INFO:ThreadId 6422] [0m[2018-01-08 19:40:14 CET] 
[36mGET[0m
    [37m:>[0m api
    [37m:>[0m settings
    [37m:>[0m sync
    [37m:>[0m progress
   [37mStatus:[0m [32mOK[0m 0.000507s[37m > [0mprogress=461612/461612 peers=0
</pre>

**For 3 more minutes the log shows:**
<pre>
[94m[node.wallet.servant:INFO:ThreadId 8644] [0m[2018-01-08 19:43:21 CET] 
[36mGET[0m
    [37m:>[0m api
    [37m:>[0m settings
    [37m:>[0m sync
    [37m:>[0m progress
   [37mStatus:[0m [32mOK[0m 0.000548s[37m > [0mprogress=461612/461612 peers=0
</pre>

**But then at 19:43:22 it is back again and shows a new total number of blocks:**
<pre>
[94m[node.wallet.servant:INFO:ThreadId 8668] [0m[2018-01-08 19:43:22 CET] 
[36mGET[0m
    [37m:>[0m api
    [37m:>[0m settings
    [37m:>[0m sync
    [37m:>[0m progress
   [37mStatus:[0m [32mOK[0m 0.002369s[37m > [0mprogress=461612/461637 peers=0
</pre>

Apparently,  the MacOS node is capable of getting back on track after an Internet outage although in the logs there is no mention of a loss of connection or a reconnect after an outage. My Linux node is not showing this behavior. Currently, it is still stuck after a restart of my router at 05:00.

I would really appreciate a comment of one of the developers on this behavior...
